docs/how-to/rccl-usage-tips.rst

```diff
@@ -82,22 +82,174 @@ set the HSA environment variable as follows:
+
+   mpirun -np 64 --bind-to numa rccl-tests/build/all_reduce_perf -b 8 -e 1G -f 2 -g 1
+
+To run RCCL allreduce on 8 GPUs in the same OAM with CPX+NPS4 mode on the MI300X, use this example:
+
+.. code-block:: shell
+
+   export ROCR_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
+
+   mpirun -np 8 --bind-to numa rccl-tests/build/all_reduce_perf -b 8 -e 1G -f 2 -g 1
+
+RCCL delivers improved allreduce performance in CPX mode for TP=8 (8 GPUs in
+the same OAM) on the MI300X.
+
+.. code-block:: shell
+
+   export HIP_FORCE_DEV_KERNARG=1
+   export RCCL_MSCCLPP_THRESHOLD=1073741824
+
+   export MSCCLPP_READ_ALLRED=1 
+   export ROCR_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
+
+   mpirun -np 8 --bind-to numa rccl-tests/build/all_reduce_perf -b 32 -e 1G -f 2 -g 1 -G 2 -w 20 -n 50
+
+Here are the benchmark results for in-place (where the output buffer is used as the input
+buffer) and out-of-place reduce bus bandwidth.
+
+.. figure:: ../data/how-to/rccl-usage-tips/in-place_reduce.png
+    :alt: In-place reduce benchmark results
+    :align: center
+
+.. figure:: ../data/how-to/rccl-usage-tips/out-of-place_reduce.png
+    :alt: Out-of-place reduce benchmark results
+    :align: center
+
+Using RCCL and CPX in PyTorch
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The PyTorch all_reduce benchmark is used to reproduce the performance reported
+by RCCL-Tests with the RCCL and CPX optimizations.
+
+.. note::
+
+   To use RCCL with CPX mode in PyTorch, check the RCCL version used by PyTorch.
+
+   For a virtualenv with a .whl-based PyTorch setup (such as nightly/rocm6.2),
+   this would be in 
+   ``<path-to-your-venv>/lib/<python-version>/site-packages/torch/lib/librccl.so``
+   This is the version of RCCL that is packaged as part of ROCm version 6.2.
+
+   RCCL for CPX mode was enabled in ROCm 6.3.0. To use the CPX features, replace the existing 
+   ``librccl.so`` with one from ROCm 6.3.0 or newer or from a local build of
+   the RCCL develop branch.
+
+To test the effects of RCCL on PyTorch, the `stas00 all reduce benchmark <https://github.com/stas00/ml-engineering/blob/master/network/benchmarks/all_reduce_bench.py>`_
+was used. The following command is used to run a single OAM allreduce
+benchmark:
+
+.. code-block:: shell
+
+   export ROCR_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 ; python -u -m torch.distributed.run --nproc_per_node=8 --rdzv_endpoint localhost:6000  --rdzv_backend c10d all_reduce_bench.py
+
+For better performance, the ``HIP_FORCE_DEV_KERNARG``, ``RCCL_MSCCLPP_THRESHOLD``,
+and ``TORCH_NCCL_USE_TENSOR_REGISTER_ALLOCATOR_HOOK`` environment variables are
+set during the benchmark in the following manner:
+
+.. code-block:: shell
+
+   export TORCH_NCCL_USE_TENSOR_REGISTER_ALLOCATOR_HOOK=1 ; export HIP_FORCE_DEV_KERNARG=1 ; export RCCL_MSCCLPP_THRESHOLD=$((2*1024*1024*1024)) ; export MSCCLPP_READ_ALLRED=1 ; export ROCR_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 ; python -u -m torch.distributed.run --nproc_per_node=8 --rdzv_endpoint localhost:6000  --rdzv_backend c10d all_reduce_bench.py
```

Comment:
It would be good to have the env variables one per line.

Reply:
Updated.